cpp/src/reductions/scan/ewm.cu

```diff
@@ -0,0 +1,379 @@
+                                           rmm::mr::device_memory_resource* mr)
+{
+  rmm::device_uvector<T> output(input.size(), stream);
+  rmm::device_uvector<pair_type<T>> pairs(input.size(), stream);
+  rmm::device_uvector<cudf::size_type> nullcnt(input.size(), stream);
+
+  if (input.has_nulls()) {
+    nullcnt = null_roll_up(input, stream);
+
+    auto device_view = column_device_view::create(input);
+    auto valid_it    = cudf::detail::make_validity_iterator(*device_view);
+    auto data =
+      thrust::make_zip_iterator(thrust::make_tuple(valid_it, nullcnt.begin(), input.begin<T>()));
+
+    thrust::transform_inclusive_scan(rmm::exec_policy(stream),
+                                     data,
+                                     data + input.size(),
+                                     pairs.begin(),
+                                     ewma_adjust_functor<T, true, true>{beta},
+                                     recurrence_functor<T>{});
+
+  } else {
+    thrust::transform_inclusive_scan(rmm::exec_policy(stream),
+                                     input.begin<T>(),
+                                     input.end<T>(),
+                                     pairs.begin(),
+                                     ewma_adjust_functor<T, false, true>{beta},
+                                     recurrence_functor<T>{});
+  }
+
+  // copy the second elements to the output for now
+  thrust::transform(rmm::exec_policy(stream),
+                    pairs.begin(),
+                    pairs.end(),
+                    output.begin(),
+                    [=] __device__(pair_type<T> pair) -> T { return pair.second; });
+
+  // denominator
+  if (input.has_nulls()) {
+    auto device_view = column_device_view::create(input);
+    auto valid_it    = cudf::detail::make_validity_iterator(*device_view);
+    auto data =
+      thrust::make_zip_iterator(thrust::make_tuple(valid_it, nullcnt.begin(), input.begin<T>()));
+
+    thrust::transform_inclusive_scan(rmm::exec_policy(stream),
+                                     data,
+                                     data + input.size(),
+                                     pairs.begin(),
+                                     ewma_adjust_functor<T, true, false>{beta},
+                                     recurrence_functor<T>{});
+  } else {
+    auto itr = thrust::make_counting_iterator<size_type>(0);
+    thrust::transform_inclusive_scan(rmm::exec_policy(stream),
+                                     itr,
+                                     itr + input.size(),
+                                     pairs.begin(),
+                                     ewma_adjust_functor<T, false, false>{beta},
+                                     recurrence_functor<T>{});
+  }
+
+  thrust::transform(
+    rmm::exec_policy(stream),
+    pairs.begin(),
+    pairs.end(),
+    output.begin(),
+    output.begin(),
+    [] __device__(pair_type<T> pair, T numerator) -> T { return numerator / pair.second; });
```

Comment:
Even if you did this wouldn't you still need to launch a copy to move the data from the second element of each pair into the output array?

Reply:
Why do you ever need to store the intermediate value in the pair though? AFAICT the pairs are output only for the above `transform_inclusive_scan` calls, they aren't used as the input anywhere, so instead of having the functor return the pair couldn't the functor be changed to just return `numerator / pair.second` and write directly into `output`?

Reply:
Note that you might need to zip output into the itr as well, and probably need to double-check that it's safe to read and write that way.

Reply:
I think I need just a little more help seeing what you are proposing. In this algorithm, we have the following steps
1.  Populate pairs with the correct data to compute the numerators with a prefix sum
2. Perform the prefix sum on the populated pairs
3. Move the second element of the (now prefix-summed) pairs to the output
4. Populate the pairs (reusing the previous allocation) with the correct data to compute the denominators with a prefix sum
5. Perform the prefix sum on the populated pairs
6. Take the second element of the denominators and elementwise divide the numerators located in the output array by them. 

In the current view, (1) and (2) are computed using a transform inclusive scan, after which if we want to reuse that allocation, we need to move the answers out of the memory and into the output (3). At this point (4) and (5) are done together using another transform inclusive scan. (6) is the intended result of this final transform. The pairs necessarily hold the intermediate result of the computation because that's where the 'answers' of the prefix sum land. 

It seems to be you are suggesting a formulation wherein (4), (5), and (6) all happen using one kernel launch. I don't quite follow how we'd do that given the full prefix sum must finish executing before the answers are available to copy to the output. Can you help me out with a quick example or maybe some pseudocode?

Reply:
I think you're right, there is no way to fuse these operations the way that I was hoping because you need the scan to complete before you can do the transform. So never mind!